lib/CL/clEnqueueNDRangeKernel.c

```diff
@@ -129,81 +217,212 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+           *
+           * Setting the coordinates with the smallest grain to 1,
+           * since they aren't good candidates for vectorizations anyway.
+           *
+           * Setting to 1 any coordinate, as a desperate measure.
+           */
+
+#define TRY_HALVE(coord) \
+if ((local_##coord & 1) == 0 && \
+    divide_chain(grain_##coord, local_##coord/2, global_##coord)) \
+  { \
+    local_##coord /= 2; \
+    continue; \
+  }
+
+#define TRY_LEAST_GRAIN(c1, c2, c3) \
+if (local_##c1 > 1 && grain_##c1 <= grain_##c2 && grain_##c1 <= grain_##c3) \
+  { \
+    local_##c1 = 1; \
+    continue; \
+  }
+
+#define DESPERATE_CASE(coord) \
+if (local_##coord > 1) \
+  { \
+    local_##coord = 1; \
+    continue; \
+  }
+          /* Halving attempt first */
+          TRY_HALVE(z) else TRY_HALVE(y) else TRY_HALVE(x)
+
+          /* Ok no luck. Find the coordinate with the smallest grain and
+           * kill that */
+          TRY_LEAST_GRAIN(z, x, y) else
+          TRY_LEAST_GRAIN(y, z, x) else
+          TRY_LEAST_GRAIN(x, y, z)
+
+          /* No luck either? Give up, kill everything */
+          DESPERATE_CASE(z) else DESPERATE_CASE(y) else DESPERATE_CASE(x)
+#undef DESPERATE_CASE
+#undef TRY_LEAST_GRAIN
+#undef TRY_HALVE
+        }
 
-        /* Finally, start setting them to zero starting from the Z
-           dimension. */
-        if (local_z > 1)
-          {
-            local_z = 1;
-            continue;
-          }
-        else if (local_y > 1)
-          {
-            local_y = 1;
-            continue;
-          }
-        else if (local_x > 1)
-          {
-            local_x = 1;
+      /* We now have the largest possible local work-group size that satisfies
+       * all the hard constraints (divide global, per-dimension bound, overall
+       * bound) and our soft constraint of being as close as possible a multiple
+       * of the preferred work-group size multiple. Such a greedy algorithm
+       * minimizes the total number of work-groups. In moderate-sized launch
+       * grid, this may result in less work-groups than the number of Compute
+       * Units, with a resulting imbalance in the workload distribution.
+       * At the same time, we want to avoid too many work-groups, since some
+       * devices are penalized by such fragmentation.
+       * Finding a good balance between the two is a hard problem, and generally
+       * depends on the device as well as the kernel utilization of its
+       * resources.
+       * Lacking that, as a first step we will simply try to guarantee that we
+       * have at least one work-group per CU, as long as the local work size
+       * does not drop below a given threshold.
+       */
+
+      /* Pick a minimum work-group size of 4 times the preferred work-group size
+       * multiple, under the assumption that this would be a good candidate
+       * below which a Compute Unit will not do enough work. */
+      const size_t min_group_size = 4*preferred_wg_multiple;
```

Comment:
Please check the GNU C style. There are some missing white spaces before and after operators and before (.

https://www.gnu.org/prep/standards/html_node/Formatting.html#Formatting